Add teardown to reset_connection at MysqlTypeLookupTest

yahonda opened the pull request:
This issue addresses one of random failure at `unlock-minitest` branch
https://travis-ci.org/rails/rails/jobs/239950090

### Summary
Restore `emulate_booleans` value every test at `Mysql2BooleanTest`
to avoid random failures at `Mysql2BooleanTest#test_column_type_without_emulated_booleans`

It is likely due to `Mysql2BooleanTest` changes `emulate_booleans` value to non default one but does not restore to original value in every tests.

### Steps to reproduce
1. Install Vagrant
2. At your host 
```
git clone https://github.com/rails/rails-dev-box.git
cd rails-dev-box
vagrant up
vagrant ssh
```

3. At rails-dev-box
```
git clone https://github.com/rails/rails.git
cd rails/activerecord
git checkout unlock-minitest
bundle install
wget https://gist.githubusercontent.com/yahonda/a41e73ee108a2d2e22006d33a74205eb/raw/8723b8acc661ae0380d5369bfca6844b3b1817bb/rep_mysql2_booleans_unlock_minitest.sh
sh ./rep_mysql2_booleans_unlock_minitest.sh
```

### Expected behavior
All unit tests should pass.

### Actual behavior
```ruby
Using mysql2
Run options: --seed 59459

# Running:

.........F..

Finished in 0.762664s, 15.7343 runs/s, 41.9582 assertions/s.

  1) Failure:
Mysql2BooleanTest#test_column_type_without_emulated_booleans [/home/ubuntu/rails/activerecord/test/cases/adapters/mysql2/boolean_test.rb:37]:
Expected: :integer
  Actual: :boolean

12 runs, 32 assertions, 1 failures, 0 errors, 0 skips
```

### System configuration
**Rails version**:unlock-minitest branch

**Ruby version**:ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux-gnu]

yahonda commented:
This pull request is for `unlock-minitest` branch at #29271

matthewd commented:
> It is likely due to Mysql2BooleanTest changes emulate_booleans value to non default one but does not restore to original value in every tests.

Isn't that exactly what the teardown is/was doing? :confused:

yahonda commented:
Yes, I might have misunderstood the cause of this failure.

I have created a minimum test case here:

```ruby
 ARCONN=mysql2 bin/test $(ls "test/cases/connection_adapters/mysql_type_lookup_test.rb" "test/cases/adapters/mysql2/boolean_test.rb") -n "/^(?:ActiveRecord::ConnectionAdapters::MysqlTypeLookupTest#(?:test_enum_type_with_value_matching_other_type|test_boolean_types)|Mysql2BooleanTest#(?:test_column_type_without_emulated_booleans))$/" --seed 41495 -v
```

It fails when `Mysql2BooleanTest#test_column_type_without_emulated_booleans` executed after 
two tests of `MysqlTypeLookupTest`.  Then I'll need to take a look at `MysqlTypeLookupTest`.

```ruby
$ ARCONN=mysql2 bin/test $(ls "test/cases/connection_adapters/mysql_type_lookup_test.rb" "test/cases/adapters/mysql2/boolean_test.rb") -n "/^(?:ActiveRecord::ConnectionAdapters::MysqlTypeLookupTest#(?:test_enum_type_with_value_matching_other_type|test_boolean_types)|Mysql2BooleanTest#(?:test_column_type_without_emulated_booleans))$/" --seed 41495 -v
Using mysql2
Run options: -n "/^(?:ActiveRecord::ConnectionAdapters::MysqlTypeLookupTest#(?:test_enum_type_with_value_matching_other_type|test_boolean_types)|Mysql2BooleanTest#(?:test_column_type_without_emulated_booleans))$/" --seed 41495 -v

# Running:

ActiveRecord::ConnectionAdapters::MysqlTypeLookupTest#test_boolean_types = 0.00 s = .
ActiveRecord::ConnectionAdapters::MysqlTypeLookupTest#test_enum_type_with_value_matching_other_type = 0.00 s = .
Mysql2BooleanTest#test_column_type_without_emulated_booleans = 0.10 s = F


Failure:
Mysql2BooleanTest#test_column_type_without_emulated_booleans [/home/ubuntu/rails/activerecord/test/cases/adapters/mysql2/boolean_test.rb:37]:
Expected: :integer
  Actual: :boolean


bin/test test/cases/adapters/mysql2/boolean_test.rb:34


Finished in 0.108958s, 27.5337 runs/s, 36.7116 assertions/s.

3 runs, 4 assertions, 1 failures, 0 errors, 0 skips
$
```

yahonda committed:
Add teardown to reset_connection at MysqlTypeLookupTest
to address `Mysql2BooleanTest#test_column_type_without_emulated_booleans`
failure

yahonda commented:
Updated the pull request to add teardown to reset_connection at MysqlTypeLookupTest including the subject changes.
